traffic_ops/traffic_ops_golang/role/roles.go

```diff
@@ -273,3 +291,522 @@ RETURNING id, last_updated`
+	//		return
+	//	}
+	//	if err != nil {
+	//		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, err)
+	//		return
+	//	}
+	//
+	//	if !api.IsUnmodified(r.Header, *existingLastUpdated) {
+	//		api.HandleErr(w, r, tx, http.StatusPreconditionFailed, api.ResourceModifiedError, nil)
+	//		return
+	//	}
+	//
+	//	rows, err := tx.Query(updateLegacyRoleQuery(), roleName, roleDesc, roleID)
+	//	if err != nil {
+	//		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, errors.New("updating role here: "+err.Error()))
+	//		return
+	//	}
+	//	rows.Close()
+	//}
+
+	if roleCapabilities != nil && *roleCapabilities != nil {
+		userErr, sysErr, errCode = deleteRoleCapabilityAssociations(inf.Tx, roleID)
+		if userErr != nil || sysErr != nil {
+			api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+			return
+		}
+		userErr, sysErr, errCode = createRoleCapabilityAssociations(inf.Tx, roleID, roleCapabilities)
+		if userErr != nil || sysErr != nil {
+			api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+			return
+		}
+	}
+	alerts := tc.CreateAlerts(tc.SuccessLevel, "role was updated.")
+	var roleResponse interface{}
+	if version.Major >= 5 {
+		var capabilities []string
+		if roleCapabilities != nil {
+			capabilities = *roleCapabilities
+		}
+		roleResponse = tc.RoleV50{
+			Name:        util.StrPtr(roleName),
+			Permissions: capabilities,
+			Description: util.StrPtr(roleDesc),
+		}
+	} else {
+		roleResponse = tc.Role{
+			RoleV11: tc.RoleV11{
+				ID:          util.IntPtr(roleID),
+				Name:        util.StrPtr(roleName),
+				Description: util.StrPtr(roleDesc),
+				PrivLevel:   util.IntPtr(privLevel),
+			},
+			Capabilities: roleCapabilities,
+		}
+	}
+	api.WriteAlertsObj(w, r, http.StatusOK, alerts, roleResponse)
+}
+
+func deleteRoleQuery() string {
+	return `DELETE FROM role WHERE id = $1`
+}
+
+func readQuery() string {
+	return `SELECT
+id,
+name,
+description,
+priv_level,
+last_updated,
+ARRAY(SELECT rc.cap_name FROM role_capability AS rc WHERE rc.role_id=id) AS permissions
+FROM role`
+}
+
+func createQuery() string {
+	return `INSERT INTO role (
+name,
+description,
+priv_level
+) VALUES (
+$1,
+$2,
+$3
+)
+RETURNING id, last_updated`
+}
+
+func createRoleCapabilityAssociations(tx *sqlx.Tx, roleID int, permissions *[]string) (error, error, int) {
+	result, err := tx.Exec(associateCapabilities(), roleID, pq.Array(permissions))
+	if err != nil {
+		return nil, errors.New("creating role capabilities: " + err.Error()), http.StatusInternalServerError
+	}
+
+	if rows, err := result.RowsAffected(); err != nil {
+		log.Errorf("could not check result after inserting role_capability relations: %v", err)
+	} else if expected := len(*permissions); int(rows) != expected {
+		log.Errorf("wrong number of role_capability rows created: %d expected: %d", rows, expected)
+	}
+	return nil, nil, http.StatusOK
+}
+
+func Validate(tx *sqlx.Tx, role tc.Role, roleV50 tc.RoleV50, version *api.Version) error {
```

Comment:
Does this need to be exported out of the package?

Instead of using different arguments based on the API version, could we not just have a different validation func for each version of a Role? I think what you're trying to do here is something like
```typescript
export function Validate(tx: sqlx.Tx, role: tc.Role | tc.RoleV50, version: api.Version) {
```
but the reality is that things like that don't work well in Go.